### PR TITLE
[kuka_ros2_control_support] Fix ros2 control support test

### PR DESCRIPTION
**Pull request by muritane**
without pre-commit currently

### CODE REVIEW

File: kuka_ros2_control_support/launch/test_bringup.launch.py [hunk 1]
```diff
@@ -100,7 +100,6 @@ def generate_launch_description():
                 "kr150_2_macro.xacro",
                 "kr150r3100_2_macro.xacro",
                 "kr210l150_macro.xacro",
-                "kr210r3100_macro.xacro",
```
Comment (gwalck) [inline kuka_ros2_control_support/launch/test_bringup.launch.py:103]: [another branch](https://github.com/StoglRobotics-forks/kuka_experimental/tree/adding-kuka210r3100-robot-foxy) adds this robot, so depending which one is merged first maybe one should not delete, but just comment out this robot ?

File: kuka_ros2_control_support/launch/bringup.launch.py [hunk 2]
```diff
@@ -162,7 +162,7 @@ def generate_launch_description():
             "rviz_file",
             default_value="moveit.rviz",
```
Comment (gwalck) [inline kuka_ros2_control_support/launch/bringup.launch.py]: this default file `moveit.rviz` does not exist, the test replaces it with [view_robot.rviz](https://github.com/StoglRobotics-forks/kuka_experimental/blob/aebb9149b5bd4902e2306d8269343fa7bd7a4318/kuka_resources/config/view_robot.rviz) which is fine, but shouldn't the default file exist at least ?

File: kuka_ros2_control_support/launch/test_bringup.launch.py (hunk 1 above)
Comment (gwalck) [inline kuka_ros2_control_support/launch/test_bringup.launch.py:103]: same

File: kuka_ros2_control_support/test/ros2_control_support.py [hunk 3]
```diff
@@ -164,10 +172,10 @@ def parse_args():
             PathJoinSubstitution([FindPackageShare(package_name), "launch", launch_file_name])
         ),
         launch_arguments={
-            "robot_description_package": launch_args["robot_description_package"],
-            "robot_description_macro_file": launch_args["robot_description_macro_file"],
-            "robot_name": launch_args["robot_name"],
```
Comment (gwalck) [inline kuka_ros2_control_support/test/ros2_control_support.py:169]: is it intentional the the `robot_name` is not passed anymore although it is still available in the combination generation? It is used in the `robot_description` (xacro) and might be needed in future checks that read the `robot_name` form the URDF?

File: kuka_ros2_control_support/urdf/common_kuka.xacro [hunk 4]
```diff
@@ -100,9 +100,6 @@
     <xacro:if value="${description_macro_file == 'kr210l150_macro.xacro'}">
       <xacro:kuka_kr210l150 prefix="$(arg prefix)"/>
     </xacro:if>
-    <xacro:if value="${description_macro_file == 'kr210r3100_macro.xacro'}">
-      <xacro:kuka_kr210r3100 prefix="$(arg prefix)"/>
-    </xacro:if>
```
Comment (gwalck) [inline kuka_ros2_control_support/urdf/common_kuka.xacro:105]: same

File: kuka_ros2_control_support/test/ros2_control_support.py [hunk 5]
```diff
@@ -179,16 +187,45 @@ class RobotDriverTest(unittest.TestCase):
...
+    #
+
+    @classmethod
+    def trajectory_state_callback(cls, msg):
+        # Save the joint names and their current positions
+        cls.joint_states = dict(zip(msg.joint_names, msg.actual.positions))
+        print(f"joint_state_future done: {cls.joint_state_future.done()}")
+        cls.joint_state_future.set_result(True)
+        cls.node.get_logger().info(f"Received joint states: {cls.joint_states}")
+
```
Comment (gwalck) [inline kuka_ros2_control_support/test/ros2_control_support.py]: I think this callback is not used at all.

File: kuka_ros2_control_support/test/ros2_control_support.py [hunk 6]
```diff
@@ -179,16 +187,45 @@ class RobotDriverTest(unittest.TestCase):
...
+    @classmethod
+    def wait_for_trajectory_state(cls, time_to_wait=3.0):
+        ret, msg = wait_for_message(
+            JointTrajectoryControllerState,
+            cls.node,
+            "/position_trajectory_controller/state",
+            time_to_wait=time_to_wait,
+        )
+        cls.assertTrue(cls, ret)
```
Comment (gwalck) [inline kuka_ros2_control_support/test/ros2_control_support.py]:
```suggestion
        cls.assertTrue(ret, "failed to receive a trajectory state message")
```
The problem is, that with no message, the assertion fails but the code continues and then throw errors due to msg being None. If one protects it from failing, later code expects the joint_states to be set and fails then later.

**Review by gwalck — CHANGES_REQUESTED** [review]
I like the revised access to the possible test choices, using python API instead of subprocess.

The tests pass, but the new joint_trajectory test should be slightly improved. I did not manage to make the test fail. I tried setting joints in URDF with lower=upper but the mock component does not respect joint limits, so not possible to see if a robot has bad limits that way. 
What would be an example of bad URDF that would make the trajectory test fail ? (to see if the test can actually fail)
A bad topic name on the trajectory controller makes the wait function thrown an exception and not fail properly.

After some discussion with @destogl one should ensure that a broken setup between the kuka robot description and the ros2_controller additional description is captured by the test. Currently if a joint name is different than `joint_a#` in the robot URDF, the tests still pass but the `robot_state_publisher` does not receive correct `joint_names` and hence the robot is all white in RViz. Even if we set the `command_names` for the controllers to random names, the tests pass, because the mock interface uses those and the test relies on matching output/input of the controller_state, not on actual robot `joint_states` matching `controller_state` names.

At least wrong names in the additional ros2_control URDF and non-matching names in their param config fails.

File: kuka_ros2_control_support/launch/test_bringup.launch.py (hunk 1 above)
Comment (gwalck) [reply kuka_ros2_control_support/launch/test_bringup.launch.py:103]: Ok, this robot should be added in the test, in the other branch not here.

File: kuka_ros2_control_support/test/ros2_control_support.py (hunk 3 above)
Comment (muritane) [reply kuka_ros2_control_support/test/ros2_control_support.py:169]: robot_name seems not necessary to instantiate urdf description

File: kuka_ros2_control_support/test/ros2_control_support.py (hunk 6 above)
Comment (muritane) [reply kuka_ros2_control_support/test/ros2_control_support.py]: thanks

File: kuka_ros2_control_support/launch/bringup.launch.py (hunk 2 above)
Comment (muritane) [reply kuka_ros2_control_support/launch/bringup.launch.py]: thanks

**Review by gwalck — APPROVED** [review]

**Comment by gwalck** [comment]
@muritane you wrote "without pre-commit" currently. Do you plan to fix the pre-commit issues ? Other than that, I am happy with the changes. Thanks.

**Comment by destogl** [comment]
If you are fixing pre-commit, please also update this option: https://github.com/StoglRobotics/ros_team_workspace/pull/135

**Comment by gwalck** [comment]
@muritane , could you also make       `use_mock_hardware` to true by default in the `test_bringup.lauch.py`  https://github.com/StoglRobotics-forks/kuka_experimental/blob/c4f2d3114976976e3144da8fdfa59e4c23cdc6cf/kuka_ros2_control_support/launch/test_bringup.launch.py#L154 because current set of default values leads the manually started test_bringup  to crash (RSI `use_rsi_communication` and EKI  `use_eki_communication` are both to **false** as well leaving not `hardware_interface` instantiated at all) even with the minimum mandatory arguments given.

**Comment by muritane** [comment]
for fixing precommit there is separate pr and still wip: https://github.com/StoglRobotics-forks/kuka_experimental/pull/30

**Pull request merged**